test/irmin-pack/test_inode.ml

```diff
@@ -198,13 +234,14 @@ module Inode_permutations_generator = struct
     tree_per_steps : inode StepSetMap.t;
   }
 
-  (** [gen_step index_list] uses brute force to generate a step such that
-      [Inter.Val.index ~depth:i] maps to the ith index in the [index_list]. *)
-  let gen_step : int list -> Path.step =
+  (** [gen_step f index_list] uses brute force to generate a step such that
+      [f step (depth, i)] maps to the ith index in the [index_list]. [f] is
+      typically [fun step (depth, i) -> Inter.Val.index ~depth s = i] *)
+  let gen_step : (step -> int * int -> bool) -> int list -> Path.step =
```

Comment:
I agree with your thought that it would be better to not pass a function here, just to keep these details from leaking out. 

You can either functorize `Inode_permutations_generator` as you mentioned or, just to throw out another option, use first class modules. Your preference!

```suggestion
  let gen_step :
      (module Irmin_pack.Inode.Internal with type Val.step = Path.step) ->
      int list ->
      Path.step =
```